bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartImpl.java

```diff
@@ -350,13 +364,33 @@ private void handleDataItem(final DataItem<?> dataItem) throws GardenaException
     }
 
     @Override
-    public void onWebSocketClose() {
-        restartWebsockets();
+    public synchronized void onWebSocketClose(String id) {
+        restartWebsocket(webSockets.get(id));
     }
 
     @Override
-    public void onWebSocketError() {
-        eventListener.onError();
+    public synchronized void onWebSocketError(String id) {
+        restartWebsocket(webSockets.get(id));
+    }
+
+    private synchronized void restartWebsocket(@Nullable GardenaSmartWebSocket socket) {
+        if (socket != null && !socket.isClosing()) {
+            logger.warn("Restarting GardenaSmart Webservice ({})", socket.getSocketID());
+            socket.stop();
+            // restart after 3 seconds
+            scheduler.schedule(() -> {
```

Comment:
Is it ok to run this asynchronously, i.e. not within the synchronized call anymore? Or would a simple `Thread.sleep(3000)` be a safer choice?

Reply:
Hi Kai,
sorry for the late answer.
I did it this way, because the functions `onWebSocketError` and `onWebSocketClose` are called at the same time, if the socket connection closes unexpectedly. When only doing a `Thread.sleep` before restarting the socket, the socket would restart twice.
To overcome this problem, I did `scheduler.schedule()` to wait 3 seconds, before restarting the socket. Because this is done in a new thread, the first call to `restartWebsocket` is immediately leaving the synchronized block, so when the second call enters the synchronized block, the socket is already closed, but not yet restarted. This prevents a second restart of the websocket.

I think this is safe and it also works fine in my openhab installation for several month now.

But if this is not a good practice, there are other options to solve this problem. For example:
- Only restart the socket, when `onWebSocketClose` is called and use `onWebSocketError` for logging purposes only. This also works fine for me, because every time the connection is closed unexpectedly, both functions are called. But I am not a Java developer, so I don't know, if it is guaranteed, that `onWebSocketClose` is called every time when an error occurs...
- If we still want to be able to restart the socket on calls to `onWebSocketError` and/or `onWebSocketClose`, I could refactor the synchronized block to get rid of `scheduler.schedule()` und use `Thread.sleep(3000)` instead. I already tested this, and it works fine. The websocket is still only restarted once.

What do you prefer?

Reply:
See my comments today about the logger level. Personally I would suggest you do something like the following..

```
    private @Nullable ScheduledFuture<?> restartTask;

    @Override
    public void onWebSocketClose() {
        // do your logging
    ...
        // do your other stuff
    ...
        scheduleRestart();
    }

    @Override
    public void onWebSocketError() {
        // do your logging
    ...
        // do your other stuff
    ...
        scheduleRestart();
    }

    private synchronized void scheduleRestart() {
        ScheduledFuture<?> restartTask = this.restartTask;
        if ((restartTask == null) || restartTask.isDone()) {
            restartTask = scheduler.schedule(() -> { restartWebsockets(); }, 3, TimeUnit.SECONDS);
         }
    }

    @Override
    public synchronized void restartWebsockets() {
        // your restart code here
        this.restartTask = null;
    }

```